src/UserSpaceInstrumentation/RegisterState.cpp

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "UserSpaceInstrumentation/RegisterState.h"
+
+#include <cpuid.h>
+#include <elf.h>
+#include <errno.h>
+#include <sys/ptrace.h>
+#include <sys/uio.h>
+
+#include <string>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/SafeStrerror.h"
+#include "absl/strings/str_format.h"
+
+namespace orbit_user_space_instrumentation {
+
+namespace {
+
+// Return the size of the XSave area on this cpu.
+[[nodiscard]] ErrorMessageOr<size_t> GetXSaveAreaSize() {
+  uint32_t eax = 0;
+  uint32_t ebx = 0;
+  uint32_t ecx = 0;
+  uint32_t edx = 0;
+  if (!__get_cpuid(1, &eax, &ebx, &ecx, &edx) || !(ecx & bit_XSAVE)) {
+    return ErrorMessage("XSAVE is not supported by Cpu.");
+  }
+  if (!__get_cpuid_count(0x0d, 0x00, &eax, &ebx, &ecx, &edx)) {
+    return ErrorMessage("XSAVE is not supported by Cpu.");
+  }
+  return static_cast<size_t>(ecx);
+}
+
+// Return offset of the YMMx registers inside the extended section of the XSave area.
+[[nodiscard]] ErrorMessageOr<size_t> GetAvxOffset() {
+  uint32_t eax = 0;
+  uint32_t ebx = 0;
+  uint32_t ecx = 0;
+  uint32_t edx = 0;
+  if (!__get_cpuid(0x01, &eax, &ebx, &ecx, &edx) || !(ecx & bit_AVX)) {
+    return ErrorMessage("AVX is not supported by Cpu.");
+  }
+  if (!__get_cpuid_count(0x0d, 0x02, &eax, &ebx, &ecx, &edx)) {
+    return ErrorMessage("AVX offset query failed.");
+  }
+  return static_cast<size_t>(ebx);
+}
+
+}  // namespace
+
+bool RegisterState::Hasx87DataStored() {
+  return (static_cast<uint64_t>(GetXSaveHeader()->xstate_bv) &
+          static_cast<uint64_t>(XSaveHeader::StateComponents::kX87)) != 0;
+}
+
+bool RegisterState::HasSseDataStored() {
+  return (static_cast<uint64_t>(GetXSaveHeader()->xstate_bv) &
+          static_cast<uint64_t>(XSaveHeader::StateComponents::kSse)) != 0;
+}
+
+bool RegisterState::HasAvxDataStored() {
+  return (static_cast<uint64_t>(GetXSaveHeader()->xstate_bv) &
+          static_cast<uint64_t>(XSaveHeader::StateComponents::kAvx)) != 0;
+}
+
+[[nodiscard]] ErrorMessageOr<void> RegisterState::BackupRegisters(pid_t tid) {
+  tid_ = tid;
+
+  iovec iov;
+  iov.iov_base = &general_purpose_registers_;
+  iov.iov_len = sizeof(GeneralPurposeRegisters);
+  auto result = ptrace(PTRACE_GETREGSET, tid, NT_PRSTATUS, &iov);
+  if (result == -1) {
+    return ErrorMessage(absl::StrFormat("PTRACE_GETREGS, NT_PRSTATUS failed with errno: %d: \"%s\"",
+                                        errno, SafeStrerror(errno)));
```

Comment:
Isn't `SafeStrerror(errno)` enough in this message (and below)? Do you really want the exact number?

Reply:
errno might be helpful when looking into the sources where this is emitted.